From: SeongJae Park <sj@kernel.org>
Subject: mm/damon/sysfs-schemes: skip tried regions update if the scheme directory has removed
Date: Mon, 14 Nov 2022 18:29:53 +0000

A DAMON sysfs interface user can start DAMON with a scheme, remove the
sysfs directory for the scheme, and then ask updates of the scheme's tried
regions.  Because the schemes tried regions update logic doesn't aware of
the situation, it results in an invalid memory access.  Fix the bug by
checking if the scheme sysfs directory exists.

Link: https://lkml.kernel.org/r/20221114182954.4745-2-sj@kernel.org
Fixes: ("mm/damon/sysfs: implement DAMOS tried regions update command")
Signed-off-by: SeongJae Park <sj@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/damon/sysfs-schemes.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/mm/damon/sysfs-schemes.c~mm-damon-sysfs-implement-damos-tried-regions-update-command-fix
+++ a/mm/damon/sysfs-schemes.c
@@ -1227,6 +1227,11 @@ static int damon_sysfs_before_damos_appl
 			break;
 		schemes_idx++;
 	}
+
+	/* user could removed the scheme sysfs dir */
+	if (schemes_idx >= sysfs_schemes->nr)
+		return 0;
+
 	sysfs_regions = sysfs_schemes->schemes_arr[schemes_idx]->tried_regions;
 	region = damon_sysfs_scheme_region_alloc(r);
 	list_add_tail(&region->list, &sysfs_regions->regions_list);
_
